packages/flame/lib/src/effects/opacity_effect.dart

```diff
@@ -1,75 +1,75 @@
-import 'package:flame/components.dart';
-import 'package:flame/src/effects/component_effect.dart';
-import 'package:flame/src/effects/controllers/effect_controller.dart';
+import 'package:flame/effects.dart';
+import 'package:flame/src/effects/provider_interfaces.dart';
 
 /// Change the opacity of a component over time.
 ///
 /// This effect applies incremental changes to the component's opacity, and
 /// requires that any other effect or update logic applied to the same component
 /// also used incremental updates.
-class OpacityEffect extends ComponentEffect<HasPaint> {
+class OpacityEffect extends Effect with EffectTarget<OpacityProvider> {
   /// This constructor will set the opacity in relation to it's current opacity
   /// over time.
   OpacityEffect.by(
     double offset,
     super.controller, {
-    this.paintId,
+    OpacityProvider? target,
     super.onComplete,
-  }) : _opacityOffset = offset;
+  }) : _opacityOffset = offset {
+    this.target = target;
+  }
 
   /// This constructor will set the opacity to the specified opacity over time.
   factory OpacityEffect.to(
     double targetOpacity,
     EffectController controller, {
-    String? paintId,
+    OpacityProvider? target,
     void Function()? onComplete,
   }) {
     return _OpacityToEffect(
       targetOpacity,
       controller,
-      paintId: paintId,
+      target: target,
       onComplete: onComplete,
     );
   }
 
   factory OpacityEffect.fadeIn(
     EffectController controller, {
-    String? paintId,
+    OpacityProvider? target,
     void Function()? onComplete,
   }) {
     return _OpacityToEffect(
       1.0,
       controller,
-      paintId: paintId,
+      target: target,
       onComplete: onComplete,
     );
   }
 
   factory OpacityEffect.fadeOut(
     EffectController controller, {
-    String? paintId,
+    OpacityProvider? target,
     void Function()? onComplete,
   }) {
     return _OpacityToEffect(
       0.0,
       controller,
-      paintId: paintId,
+      target: target,
       onComplete: onComplete,
     );
   }
 
   double _opacityOffset;
   double _roundingError = 0.0;
-  final String? paintId;
```

Comment:
Looks like this `paintId` was buggy, so good riddance. Generally, `HasPaint` is supposed to have a `<T>` generic parameter, so that `paintId` would be of type `T?`. But the class was declared as `ComponentEffect<HasPaint>`, thus `T=Object`. But when `T=Object` the paint ids are not allowed (there is an assert for that).

Not sure how (and whether) it all worked before, but good thing we don't have to worry about that anymore.

Reply:
Yeah, I don't think most users know that `HasPaint` can actually hold multiple paints. Even our `Flower` class from examples stores multiple paints within itself, even though it has `HasPaint`  mixin 😅